processor/src/debug.rs

```diff
@@ -0,0 +1,95 @@
+use crate::{ExecutionError, Felt, Process, StarkField};
+use core::fmt;
+use vm_core::Word;
+
+/// VmState holds a current process state information at a specific clock cycle.
+#[derive(Clone, Debug, PartialEq)]
+pub struct VmState {
+    pub clk: usize,
+    pub fmp: Felt,
+    pub stack: Vec<Felt>,
+    pub memory: Vec<(u64, Word)>,
+}
+
+impl fmt::Display for VmState {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let stack: Vec<u64> = self.stack.iter().map(|x| x.as_int()).collect();
+        let memory: Vec<(u64, [u64; 4])> = self
+            .memory
+            .iter()
+            .map(|x| (x.0, word_to_ints(&x.1)))
+            .collect();
+        write!(
+            f,
+            "clk={}, fmp={}, stack={:?}, memory={:?}",
+            self.clk, self.fmp, stack, memory
+        )
+    }
+}
+
+/// Iterator that iterates through vm state at each step of the execution.
+/// This allows debuging or replaying ability to view various process state
+/// at each clock cycle.
+/// If the execution returned an error, it returns that error on the clock cycle
+/// it stopped.
+pub struct VmStateIterator {
+    process: Process,
+    error: Option<ExecutionError>,
+    clk: usize,
+}
+
+impl VmStateIterator {
+    pub(super) fn new(process: Process, result: Result<(), ExecutionError>) -> Self {
+        Self {
+            process,
+            error: result.err(),
+            clk: 0,
+        }
+    }
+}
+
+impl Iterator for VmStateIterator {
+    type Item = Result<VmState, ExecutionError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.clk > self.process.system.clk() {
+            match &self.error {
+                Some(_) => {
+                    let error = core::mem::take(&mut self.error);
+                    return Some(Err(error.unwrap()));
+                }
+                None => return None,
+            }
+        }
+
+        let mut mem_clk = self.clk;
+        if mem_clk > 0 {
+            mem_clk -= 1;
+        }
```

Comment:
I'd prefer to update the `Memory:: get_values_at()` function. It would be cleaner to encapsulate all the logic there.

I'd change the signature of that function to something like:
```Rust
pub fn get_values_at(&self, range: RangeInclusive<u64>, step: usize) -> Vec<(u64, Word)>
```
And then, since we know that memory at `clk=0` is empty, the first thing within the function could be:
```Rust
if step == 0 {
    return Vec::new();
}
```

Reply:
We do need to keep the step optional since None means the latest clk. I'll change the check to also check for the option

Reply:
I'd just move the "else" branch directly into `get_values()` method so that these methods don't rely on each other.

Reply:
ok fixed